File: framework/tests/ut/interface/src/utils/test_plugin.cpp
/**
 * Copyright (c) 2025-2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

/*!
* \file test_plugin.cpp
* \brief
*/

#include <gtest/gtest.h>
#include <string>

#include "tilefwk/function.h"
#include "tilefwk/tilefwk_op.h"
#include "tilefwk/tilefwk.h"

#include "interface/plugin/plugin.h"
#include "interface/utils/string_utils.h"

using namespace npu::tile_fwk;

TEST(PluginTest, Basic) {
    struct Compute {
        static std::string ComputeAdd(const std::string &filepath, const std::string &source) {
            return source + "Add" + filepath;
        }
        static std::string ComputeSub(const std::string &filepath, const std::string &source) {
            return source + "Sub" + filepath;
        }
    };
    PluginManager &manager = PluginManager::GetInstance();

    EXPECT_TRUE(manager.AddPluginCodegenSrc("add", Compute::ComputeAdd));
    EXPECT_FALSE(manager.AddPluginCodegenSrc("add", Compute::ComputeSub));
    EXPECT_TRUE(manager.AddPluginCodegenSrc("sub", Compute::ComputeSub));

    EXPECT_EQ(2, manager.GetPlugin<PluginCodegenSrc>().size());

    std::string code = manager.RunPluginCodegenSrc("1", "2");
    EXPECT_EQ("2Add1Sub1", code);

    manager.ClearPlugin();
    EXPECT_EQ(0, manager.GetPlugin<PluginCodegenSrc>().size());
<<<<<<< HEAD

    std::string code2 = manager.RunPluginCodegenSrc("1", "2");
    EXPECT_EQ("2", code2);
}
=======
}

TEST(PluginTest, Codegen) {
    TileShape::Current().SetVecTile(32, 32);
    TileShape::Current().SetCubeTile({32, 32}, {32, 32}, {32, 32});

    std::vector<int> opList;
    auto addLinePrefix = [&](const std::string &filepath, const std::string &source) {
        (void)filepath;
        opList.push_back(0);
        return source;
    };
    PluginManager::GetInstance().AddPluginCodegenSrc("AddPrefix", addLinePrefix);

    int n = 8;
    int s = 32;

    Tensor t0(DT_FP32, {n * s, s}, "t0");  // [32*8, 32]
    Tensor t1(DT_FP32, {n * s, s}, "t1");  // [32, 32]
    Tensor t2(DT_FP32, {s, s}, "t2");  // [32, 32]
    Tensor out(DT_FP32, {n * s, s}, "out");

    FUNCTION("main", {t0, t1, t2}, {out}) {
        Tensor s0Out;
        LOOP("L0", FunctionType::DYNAMIC_LOOP, _, LoopRange(1)) {
            (void)_;
            s0Out = Sub(t1, t0);
        }
        LOOP("L1", FunctionType::DYNAMIC_LOOP, i, LoopRange(n)) {
            Tensor t0s = View(s0Out, {s, s}, {i * s, 0});
            Tensor t3 = Add(t0s, t2);
            Assemble(t3, {i * s, 0}, out);
        }
    }
    EXPECT_EQ(2, opList.size());
    PluginManager::GetInstance().ClearPlugin();
}
>>>>>>> e93b3b6... feat(codegen): Add source_cpp ir serializer as codegen

File: python/tests/st/machine/test_plugin.py
#!/usr/bin/env python3
# coding: utf-8
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------

import re
import os
import sys

import pypto

import torch
import torch_npu

def add2sub(filepath:str, source:str) -> str:
    print(source, filepath)
    return source

@pypto.jit(codegen_src_plugins=[pypto.Plugin('add2sub', add2sub)])
def cust_dyn_func_add(a, b, c, tiling=32):
    pypto.set_vec_tile_shapes(tiling, tiling)
    for _ in pypto.loop(1, name="s0", idx_name="k"):
        c[:] = pypto.add(a, b)


def test_codegen_src_plugin():
    device_id = os.environ.get('TILE_FWK_DEVICE_ID', 0)
    torch.npu.set_device(int(device_id))

    tiling = 32
    n, m = tiling * 1, tiling * 1

    # prepare data
    a_rawdata = torch.ones((n, m)) * 10
    a_data = a_rawdata.to(dtype=torch.int32, device=f'npu:{device_id}')

    b_rawdata = torch.ones((n, m))
    b_data = b_rawdata.to(dtype=torch.int32, device=f'npu:{device_id}')

    c_data = torch.zeros((n, m), dtype=torch.int32, device=f'npu:{device_id}')

    # def inputs and outputs
    inputs = [a_data, b_data]
    outputs = [c_data]
    pto_inputs = [pypto.from_torch(tensor, f"IN_{idx}") for idx, tensor in enumerate(inputs)]
    pto_outputs = [pypto.from_torch(tensor, f"OUT_{idx}") for idx, tensor in enumerate(outputs)]

    cust_dyn_func_add(pto_inputs[0], pto_inputs[1], pto_outputs[0], tiling)
    torch_npu.npu.synchronize()

    golden = torch.ones((n, m)) * 9
    assert torch.allclose(golden.int(), c_data.cpu(), atol=1e-5)

if __name__ == '__main__':
    test_codegen_src_plugin()